p2p/shred.md

```diff
@@ -0,0 +1,313 @@
+Identifies the network fork of the block that this shred is part of.
+
+### Data Shred Header (rev. v2)
+
+The data shred header
+
+Offsets are relative to the start of the common header.
+
+| Offset | Size | Type  | Name           | Purpose                       |
+|--------|-----:|-------|----------------|-------------------------------|
+| `0x53` |   4B | `u32` | `parent_dist`  | Slot distance to parent block |
+| `0x55` |   1B | `u8`  | `data_flags`   | Data Flags                    |
+| `0x56` |   2B | `u16` | `size`         | Total Size                    |
+| `0x58` |    - | bytes | `data_payload` | Data Payload                  |
+
+The data flags contain the following fields. (LSB 0 numbering)
+
+| Bits  | Type | Name             | Purpose            |
+|-------|------|------------------|--------------------|
+| `7`   | bool | `block_complete` | Block complete bit |
+| `6`   | bool | `batch_complete` | Batch complete bit |
+| `0:6` | `u6` | `batch_tick`     | Batch tick number  |
+
+#### Field: Block complete bit
+
+Set to one if this data shred is the last shred for this block.
+Otherwise, set to zero.
+
+The batch complete bit must also be set if set to one.
+
+#### Field: Batch complete bit
+
+Set to zero if the next immediate shred is part of the same batch.
+
+Otherwise, set to one to mark the end of the current batch.
+
+### Data Shred Header (rev. v1)
+
+Revision v1 data shreds.
+
+The total size field is hardcoded to 1228 bytes.
+
+| Offset | Size | Type  | Name           |  Purpose                       |
+|--------|-----:|-------|----------------|--------------------------------|
+| `0x53` |   4B | `u32` | `parent_dist`  |  Slot distance to parent block |
+| `0x55` |   1B | `u8`  | `data_flags`   |  Data Flags                    |
+| `0x56` |    - | bytes | `data_payload` |  Data Payload                  |
+
+## Code Shreds
+
+## Legacy Authentication
+
+In the legacy authentication scheme, the block producer signs the content of each shred.
+
+The message to be signed begins 64 bytes past the beginning of the common shred header (skipping the signature field) and includes the entire rest of the shred.
+The resulting signature is placed into the _block producer signature_ field.
+
+## Merkle Authentication
+
+In the Merkle authentication scheme, the block producer signs the Merkle root over the FEC set of shreds.
+Consequently, the signature field carries identical content for all shreds in the same FEC set.
+
+The Merkle node size is truncated to 20 bytes.
+
+## Shred Construction
+
+Block producers create and broadcast shreds to the validator network while active.
+
+This occurs as a streaming process wherein shreds are constructed in the earliest possible opportunity.
+
+The construction of shreds requires the following steps:
+1. Block Entry Batching
+2. Data Shredding
+3. Erasure Coding
+4. Signing
+
+## Block Entry Batching
+
+Batching creates sub-groups of block entries, each of which are serialized into one byte array.
```

Comment:
it's worth noting that entries are non-contentious batches of txs, this way clients are inclined to make this assumption